### ISSUE TITLE
Таймер учета времени

### ISSUE THREAD

**Issue by Tahallus** [BlizD/Tasks#21]
Предлагаю сделать таймер для подсчета времени на реализацию задачи.
Когда-то делал таймер для себя простой на видео демонстарция как работает.
https://youtu.be/KJ6nc2_Cj1w

**Comment by BlizD** [comment]
Привет. 
Спасибо, за предложение. 
Сейчас, можете зайти в задачу и на закладке "Учет времени" если нажать на кнопку "Открытие" в строке. Будет в целом тот же эффект, что и от таймера о котором Вы говорите.

**Comment by BlizD** [comment]
![_ _ _ _](https://cloud.githubusercontent.com/assets/10989306/21642672/03209952-d295-11e6-89cd-d693cbd53ec0.gif)

**Comment by Tahallus** [comment]
Не очень удобно если срочно надо переключиться между задачами, а потом вернуться к прошлой задаче. и получается надо не забывать нажать на открыть, и задачу записывать постоянно чтобы случайно не закрыть и время не потерять.

**Comment by BlizD** [comment]
А в Вашем случае если надо переключится на выполнение другой задачи, то что надо сделать?
Можете по шагово расписать?

В текущем варианте надо выполнить следующее:
1. открыть задачу;
1. на закладке "Учет времени" нажать на кнопку "Открытие" или руками указать датувремя;
1. Записать задачу

**Comment by BlizD** [comment]
Понятно, а если смотреть для текущей конфигурации, то как это будет выглядеть? Основная работа с задачами происходит или в канбан доске или в справочнике задачи. Добавлять еще формы для задач пока не надо.

**Comment by Tahallus** [comment]
http://dropmefiles.com/r1u34
Вот что быстро сделал.

**Comment by BlizD** [comment]
Вова, 
пожалуйста поймите меня правильно, не воспринимайте как критику.
Но перед тем как, что то реализовывать новое в базе, как минимум надо согласовать изменения добавление метаданных и прочего и вообще в целом согласовать идею.

Посмотрел CF, что не устраивает:
- затраченное время (фактическое) сейчас хранится в ТЧ: ФактическиеЧасы, поэтому добавлять новый еще один регистр сведений не верно;
- задачи на паузе хранить в отдельном регистре сведений неверно. Надо просто проставить ДатуОкончания в ТЧ: ФактическиеЧасы;
- справочник "Партнеры" для данной доработки не нужен;
- в справочнике "Задачи" в форме списка, надо вместе подумать как лучше бы организовать 
запуск/остановку таймера. Два табличных поля для организации таймера, на мой взгляд это слишком много места занимает. Плюс не понятно как запускать/останавливать таймера в Канбан доске. 

Если все таки еще есть желание на доработку конфигурации, то готов вести конструктивную беседу.

**Comment by BlizD** [comment]
Раз мы гитхабе, тогда надо использовать его возможности. Необходимо только соблюсти вот такие правила:
1. Использовать платформу 8.3.9.1850
1. Сделайте fork проекта с ветки "develope";
1. Получите свою локальную копию;
1. Выполняете задачу на своей локальной копии;
1. Далее сделайте pull-request в проект. https://github.com/BlizD/Tasks/pulls;
ВАЖНО: выбирайте ветку features, а не master или develope;
ВАЖНО: убедитесь, что Ваш fork синхронизирован с оригинальным репозиторием develope

**Comment by Tahallus** [comment]
Эту поделку я делал примерно 30 минут, только чтобы показать принцип, так что тут много что можно критиковать.
1. По сути РС это все таже табличная часть, просто постоянно изменять задачу думаю не очень.
2. Когда задача ставить на паузу то в РС Затраченное время дата фиксируется, РС ЗадачиНаПаузе чисто чтобы отражать в интерфейсе.
3. Ну смотря для кого ориентир, у вас это внутренние задачи компании, а кому-то нужны в разрезе клиентов, это делал чисто для себя, нужно или нет это вас решать
4. Опять же я отправил код не с целью включения в релиз, а как демонстрация концепции.

На счет совместной разработки я понял все же цели у нас разные, учитывая что вы пишите что справочник Партнеры не нужны то цель ваша это учет задач внутри компании, мне больше нужна конфа для фрилансера с учетом партнеров взаиморасчетов

**Comment by BlizD** [comment]
>По сути РС это все таже табличная часть, просто постоянно изменять задачу думаю не очень.

Да, Вы, правы в РС тоже можно хранить фактические часы. Но сейчас фактические часы хранятся в табличной части в справочнике "Задачи". По поводу плохо ли записывать задачу каждый раз, по моему мнению это не плохо. 

>Когда задача ставить на паузу то в РС Затраченное время дата фиксируется, РС ЗадачиНаПаузе чисто чтобы отражать в интерфейсе.

Нужна ли информация о списке задач которые стоят на паузе? Мне кажется пока нет такой необходимости. Если для задачи не запущен таймер, тогда она в паузе или уже сделана.

>Ну смотря для кого ориентир, у вас это внутренние задачи компании, а кому-то нужны в разрезе клиентов, это делал чисто для себя, нужно или нет это вам решать

В данный момент чтобы задачи делить по клиентам, можно добавить доп. реквизит в режиме исполнения. И по нему можно делить задачи и делать отборы в отчетах и канбан доске.

>Опять же я отправил код не с целью включения в релиз, а как демонстрация концепции.

Если цель была лишь поделится идеей, тогда идею понял, спасибо.

>На счет совместной разработки я понял все же цели у нас разные, учитывая что вы пишите что справочник Партнеры не нужны то цель ваша это учет задач внутри компании, мне больше нужна конфа для фрилансера с учетом партнеров взаиморасчетов

Мне сложно оценить потребность такой доработки, т.к. взаиморасчеты обычно ведутся в УТ, БП, УПП, ERP и других типовых программах. Мне казалось, что от данной конфигурации достаточно выгрузить список задач и часы которые надо выставить клиенту. Типа акта выполненных работ. 
В целом же, что касается "Партнеров" учета взаиморасчетов, актов, счетов я не говорю, что их никогда не буду добавлять в конфигурацию. Просто мне кажется это мало кому может потребоваться.

**Comment by relieverz** [comment]
Ребята, поддерживаю ТС с данной задачей, для крупных команд необходим механизм учета трудозатрат. В таких случаях необходимо понимать, сколько было потрачено времени всей команды на выполнение проекта. Да и могут на основе этого расчитывать заработную плату.

**Comment by BlizD** [comment]
@relieverz Добрый день, Антон. 
>Ребята, поддерживаю ТС с данной задачей, для крупных команд необходим механизм учета трудозатрат. В таких случаях необходимо понимать, сколько было потрачено времени всей команды на выполнение проекта. Да и могут на основе этого расчитывать заработную плату.

Пока таких планов нет. Если будет желание доработать готов обсудить и помочь.

**Comment by capitoshko** [comment]
Таймер еще актуален? кнопание по дате календаря такое себе решение...

**Comment by CptAFK** [comment]
Есть пару идей как можно в канбан это припилить, попробую в выходные натыкать

**Comment by BlizD** [comment]
@CptAFK в этом направлении уже что у себя делал @capitoshko. Попробуйте узнать какие он идеи реализовал.

**Comment by CptAFK** [comment]
МЕТАДАННЫЕ:
Реквизит формы: СтруктураТаймер (произвольный)
ОбщийМакет.узСловарьПоУмолчанию добавил строку для новой кнопки контекстного меню

![Таймер](https://user-images.githubusercontent.com/6850355/59103662-e2f70180-8948-11e9-87ad-4d0f34b0a378.gif)

**Comment by BlizD** [comment]
@CptAFK классно выглядит. А cf можешь скинуть?

**Comment by CptAFK** [comment]
накачу изменения на чистую базу и скину, сегодня попозже, или завтра в течении дня

**Comment by CptAFK** [comment]
@BlizD Накачу изменения на чистую базу и скину, сегодня попозже, или завтра в течении дня.
Код возможно не идеальный, я только учусь, готов исправляться))) В разработке 1С всего около 2х лет, 90% задач на обычных формах было.

**Comment by capitoshko** [comment]
Мы добавили 2 кнопки. Запустить таймер, остановить таймер. При нажатии кнопки Остановить открываентся форма ввода факта работ. Если факт введен - можно выбирать новую задачу для таймера

**Comment by 79vlad** [comment]
@CptAFK  & @capitoshko  Было бы не плохо смиксовать Ваши наработки: запускать/останавливать как у @CptAFK , а обрабатывать остановку как у @capitoshko  (у нас немного доработана форма факта и там заполняются (из задачи) дополнительные данные с возможностью изменения)

**Comment by CptAFK** [comment]
Надо поэксперементировать, думаю стоит добавить флаг, отображать форму ввода или нет, т.к. я на мелкие задачи бью все, в иерархии, мне описание по выполненным работать не требуется. Можно конечно доработать еще сильнее,  я сейчас каждый раз документ факта работ создаю, что вероятно не очень правильно, т.к. в факт работ можно строки в тч добавлять. (Только об этом подумал, пока сообщение писал)

**Comment by CptAFK** [comment]
Можно так, при переключении задачи/отключении таймера он выводит окно ввода факта работа, если этот факт работ записать, он в следующий раз добавит время таймера к этом документу, если провести, в следующий раз будет создан новый документ.
